File: python_files/raylib/nature_of_code/random_walker.py
import pyray as p
import random

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
SCREEN_TITLE = "Random Walker"
SCREEN_BACKGROUND = p.BLACK
GAME_FPS = 60


def main():
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)
    p.set_target_fps(GAME_FPS)

    width = 6
    height = 6
    color = p.WHITE

    lst = [p.Vector2(p.get_screen_width() / 2, p.get_screen_height() / 2)]

    active = True

    while not p.window_should_close():
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            if active:
                active = False
            elif not active:
                active = True

        if active:
            pos = p.Vector2(lst[-1].x, lst[-1].y)
            choice = random.randint(1, 4)

            if choice == 1:
                pos.x += width
            elif choice == 2:
                pos.x -= width
            elif choice == 3:
                pos.y += width
            elif choice == 4:
                pos.y -= width

            if pos.x <= 0:
                pos.x = 0
            elif pos.x >= p.get_screen_width() - width:
                pos.x = p.get_screen_width() - width
            elif pos.y <= 0:
                pos.y = 0
            elif pos.y >= p.get_screen_height() - height:
                pos.y = p.get_screen_height() - height

            lst.append(pos)

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        for i in lst:
            p.draw_rectangle_rec((i.x, i.y, width, height), color)

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: c_files/raylib/nature_of_code/random_walker.c
#include "raylib.h"
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    const int screenWidth = 800;
    const int screenHeight = 600;
    const char screenTitle[] = "";
    const Color screenBackground = RAYWHITE;
    const int gameFps = 60;

    InitWindow(screenWidth, screenHeight, screenTitle);
    SetTargetFPS(gameFps);

    float width = 6.0f;
    float height = 6.0f;
    Color color = BLACK;

    Vector2 *list = malloc(sizeof(Vector2));
    if (list == NULL)
    {
        printf("Memory not allocated\n");
        return 1;
    }

    list[0] = (Vector2){(float)GetScreenWidth() / 2, (float)GetScreenHeight() / 2};
    int size = 1;

    while (!WindowShouldClose())
    {
        int choice = GetRandomValue(1, 4);
        Vector2 pos = list[size - 1];

        if (choice == 1)
        {
            pos.x += 6;
        }
        else if (choice == 2)
        {
            pos.x -= 6;
        }
        else if (choice == 3)
        {
            pos.y += 6;
        }
        else if (choice == 4)
        {
            pos.y -= 6;
        }

        if (pos.x <= 0)
        {
            pos.x = 0;
        }
        else if (pos.x >= GetScreenWidth() - width)
        {
            pos.x = GetScreenWidth() - width;
        }
        else if (pos.y <= 0)
        {
            pos.y = 0;
        }
        else if (pos.y >= GetScreenHeight() - height)
        {
            pos.y = GetScreenHeight() - height;
        }

        list = realloc(list, (size + 1) * sizeof(Vector2));
        if (list == NULL)
        {
            printf("Memory not reallocated\n");
            return 1;
        }
        list[size] = pos;
        size++;

        BeginDrawing();
        ClearBackground(screenBackground);

        for (int i = 0; i < size; i++)
        {
            DrawRectangleRec((Rectangle){list[i].x, list[i].y, width, height}, color);
        }

        EndDrawing();
    }

    CloseWindow();

    return 0;
}
